indiserver/ClInfo.cpp

```diff
@@ -0,0 +1,323 @@
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+#include "ClInfo.hpp"
+#include "Msg.hpp"
+#include "DvrInfo.hpp"
+#include "Constants.hpp"
+#include "Utils.hpp"
+#include "Property.hpp"
+
+ConcurrentSet<ClInfo> ClInfo::clients;
+
+// root will be released
+void ClInfo::onMessage(XMLEle * root, std::list<int> &sharedBuffers)
+{
+    char *roottag    = tagXMLEle(root);
+
+    const char *dev  = findXMLAttValu(root, "device");
+    const char *name = findXMLAttValu(root, "name");
+    int isblob       = !strcmp(tagXMLEle(root), "setBLOBVector");
+
+    /* snag interested properties.
+     * N.B. don't open to alldevs if seen specific dev already, else
+     *   remote client connections start returning too much.
+     */
+    if (dev[0])
+    {
+        // Signature for CHAINED SERVER
+        // Not a regular client.
+        if (dev[0] == '*' && !this->props.size())
+            this->allprops = 2;
+        else
+            addDevice(dev, name, isblob);
+    }
+    else if (!strcmp(roottag, "getProperties") && !this->props.size() && this->allprops != 2)
+        this->allprops = 1;
+
+    /* snag enableBLOB -- send to remote drivers too */
+    if (!strcmp(roottag, "enableBLOB"))
+        crackBLOBHandling(dev, name, pcdataXMLEle(root));
+
+    if (!strcmp(roottag, "pingRequest"))
+    {
+        setXMLEleTag(root, "pingReply");
+
+        Msg * mp = new Msg(this, root);
+        pushMsg(mp);
+        mp->queuingDone();
+        return;
+    }
+
+    /* build a new message -- set content iff anyone cares */
+    Msg* mp = Msg::fromXml(this, root, sharedBuffers);
+    if (!mp)
+    {
+        log("Closing after malformed message\n");
+        close();
+        return;
+    }
+
+    /* send message to driver(s) responsible for dev */
+    DvrInfo::q2RDrivers(dev, mp, root);
+
+    /* JM 2016-05-18: Upstream client can be a chained INDI server. If any driver locally is snooping
+    * on any remote drivers, we should catch it and forward it to the responsible snooping driver. */
+    /* send to snooping drivers. */
+    // JM 2016-05-26: Only forward setXXX messages
+    if (!strncmp(roottag, "set", 3))
+        DvrInfo::q2SDrivers(NULL, isblob, dev, name, mp, root);
+
+    /* echo new* commands back to other clients */
+    if (!strncmp(roottag, "new", 3))
+    {
+        q2Clients(this, isblob, dev, name, mp, root);
+    }
+
+    mp->queuingDone();
+}
+
+void ClInfo::close()
+{
+    if (verbose > 0)
+        log("shut down complete - bye!\n");
+
+    delete(this);
+
+#ifdef OSX_EMBEDED_MODE
+    fprintf(stderr, "CLIENTS %d\n", clients.size());
+    fflush(stderr);
+#endif
+}
+void ClInfo::q2Clients(ClInfo *notme, int isblob, const std::string &dev, const std::string &name, Msg *mp, XMLEle *root)
+{
+    /* queue message to each interested client */
+    for (auto cpId : clients.ids())
+    {
+        auto cp = clients[cpId];
+        if (cp == nullptr) continue;
```

Comment:
This `ConcurrentSet<T>` is an odd design and it's not exactly clear what this container is doing that cannot be otherwise accomplished by using `std::map` instead.